[sparsity] WeightNormSparsifier

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* **#58955 [sparsity] WeightNormSparsifier**
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

Commit:
[sparsity] WeightNormSparsifier

[ghstack-poisoned]

vadimkantorov commented on torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:
```diff
@@ -0,0 +1,37 @@
+
+    def update_mask(self, layer, sparsity_level, sparsity_pattern, **kwargs):
+        if sparsity_level <= 0:
+            layer.mask = torch.ones(layer.weight.shape)
```
layer.mask could poptentially be on another device than `layer.weight`, right?

Commit:
Update on "[sparsity] WeightNormSparsifier"




[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"


Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:
```diff
@@ -0,0 +1,51 @@
+        elif sparsity_level >= 1.0:
+            layer.mask.data = torch.zeros(layer.weight.shape, device=layer.weight.device)
+        else:
+            ww = layer.weight * layer.weight
```
```
original_weight = layer.parametrizations.weight[0].original
ww = original_weight * original_weight
```

Commit:
Update on "[sparsity] WeightNormSparsifier"




Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"




Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"




Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"




Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"




Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"




Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

raghuramank100 commented on torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:
```diff
@@ -0,0 +1,53 @@
+                 sparsity_level=0.5, sparsity_pattern=(1, 4),
+                 zeros_per_block=None):
+        if zeros_per_block is None:
+            zeros_per_block = reduce((lambda x, y: x * y), sparsity_pattern)
```
Is this restricted to 2D weights?

z-a-f replied:
No, the reduce can take care of any dimensions.

raghuramank100 commented on torch/ao/sparsity/sparsifier/base_sparsifier.py:
```diff
@@ -177,6 +178,14 @@ def convert(self):
     @abc.abstractmethod
-    def step(self):
-        return
+    def update_mask(self, layer, **kwargs):
```
Lets call this function as step instead of update mask. The reason is that:
 For every call of forward, we want the sparsifier to be called. This is to allow for updating of any stats used for mask calculation: This could be gradients, weight norms, activations etc.
The rate at  which the mask is updated can be controlled by the scheduler. At a high level, we do this:
``` python
def step(self):
  # Always update stats
  # Depending on scheduler, call self.update_mask
```

raghuramank100 commented on torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:
```diff
@@ -0,0 +1,53 @@
+        # TODO: Add support for multiple parametrizations for the same weight
+        mask = layer.parametrizations.weight[0].mask
+        if sparsity_level <= 0:
+            mask.data = torch.ones(layer.weight.shape, device=layer.weight.device)
```
Why mask.data instead of mask here?

z-a-f replied:
This is a constraint that we must have, otherwise a new variable will be created

raghuramank100 commented on torch/ao/sparsity/sparsifier/weight_norm_sparsifier.py:
```diff
@@ -0,0 +1,53 @@
+            for row, col in zip(rows, cols):
+                new_mask[row:row + sparsity_pattern[0],
+                         col:col + sparsity_pattern[1]] = 0
+            mask.data *= new_mask
```
Why are we updating mask.data here? Why not mask=new_mask?

z-a-f replied:
We are not allowed to use non-inplace assignments. All assignments to the mask must be in-place

raghuramank100 commented on torch/ao/sparsity/sparsifier/base_sparsifier.py:
```diff
@@ -177,6 +178,14 @@ def convert(self):
+        with torch.no_grad():
+            for config in self.module_groups:
+                module = config['module']
+                self.update_mask(module, **config)
```
NIT: Since update_mask is a member of the sparsifier object, why have it take arguments?

z-a-f replied:
Oh that's because the `update_mask` is just a logic of changing generic masks. The config specifies which layers is the update called for. I am doing it so that the user would need to write minimum amount of code.

raghuramank100 commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -45,6 +52,14 @@ def test_constructor(self):
+    def test_step(self):
+        model = Model()
+        sparsifier = ImplementedSparsifier(test=3)
+        sparsifier.enable_mask_update = True
```
By default, can enable_mask_update be True, so that in the absence of a scheduler, the mask is updated with every call to step? This way we can avoid requiring users to set this attribute as an additional step.

z-a-f replied:
Part of the base sparsifier PR

raghuramank100 commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -55,3 +70,37 @@ def test_state_dict(self):
+        model = Model()
+        sparsifier = ImplementedSparsifier(test=3)
+        sparsifier.prepare(model, [model.linear])
+        assert hasattr(model.linear, 'mask')
```
This shouldnt be the case isn't it? Isnt mask an attr of the fake-sparse module?

raghuramank100 commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -55,3 +70,37 @@ def test_state_dict(self):
+            # The module_groups are unordered
+            assert g['path'] in ('seq.0', 'linear')
+
+    def test_logic(self):
```
Can we call this test_step()?

raghuramank100 commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -55,3 +70,37 @@ def test_state_dict(self):
+        sparsifier.prepare(model, config=[model.linear])
+        sparsifier.enable_mask_update = True
+        sparsifier.step()
+        self.assertAlmostEqual(model.linear.mask.mean().item(), 0.5, places=2)
```
Do we need a test to check that the masks are serialized correctly as part of the model?

z-a-f replied:
I thought those are tested when I test the parametrization state tests.

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).

Commit:
Update on "[sparsity] WeightNormSparsifier"


Implements the weight norm sparsifier.
This type of sparsifier computes the norm of the weights, sorts them, and zeroes-out the target fraction of them.

The main imeplemented method is `update_mask`, which holds the main logic of changing the masks.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970960](https://our.internmc.facebook.com/intern/diff/D28970960)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970960).
